[Layout] Fix crash in WebCore::RenderGrid::gridAreaRangeForOutOfFlow

zakariaridouh opened the pull request:
#### f647097804305b55bfdced7ba9b590202fdc16cc
<pre>
[Layout] Fix crash in WebCore::RenderGrid::gridAreaRangeForOutOfFlow
<a href="https://bugs.webkit.org/show_bug.cgi?id=300354">https://bugs.webkit.org/show_bug.cgi?id=300354</a>
<a href="https://rdar.apple.com/161639703">rdar://161639703</a>

Reviewed by Alan Baradlay.

Fix a crash in WebCore::RenderGrid::gridAreaRangeForOutOfFlow where we haven&apos;t completed
in-flow placement and grid sizing yet, and so our positions vector doesn&apos;t have valid values.

This aligns us closer with the spec, where only in flow children are
eligible for margin collapsing.

The root cause of this crash is we&apos;re attempting to do margin collapsing with an out of flow child, but
we can only do this for in flow children.

[1] &lt;<a href="https://drafts.csswg.org/css2/#collapsing-margins">https://drafts.csswg.org/css2/#collapsing-margins</a>&gt;

* LayoutTests/fast/css-grid-layout/grid-crash-area-range-for-out-of-flow-expected.txt: Added.
* LayoutTests/fast/css-grid-layout/grid-crash-area-range-for-out-of-flow-no-text-expected.txt: Added.
* LayoutTests/fast/css-grid-layout/grid-crash-area-range-for-out-of-flow-no-text.html: Added.
* LayoutTests/fast/css-grid-layout/grid-crash-area-range-for-out-of-flow.html: Added.
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::isSelfCollapsingBlock const):

Canonical link: <a href="https://commits.webkit.org/301976@main">https://commits.webkit.org/301976@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cd7ae324e60fb83a1116c3fbed4c03e184146a08

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/127637 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/47285 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/38453 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/134908 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/79194 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/8f478cfb-4098-4d59-8068-4697304831bf) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/129509 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/47908 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/55816 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/97160 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/65078 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/726bb257-b304-43f9-bcc8-e1641dfa2dc1) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/130585 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/38316 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/114346 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/77641 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/ebc9e5f6-53ea-4fd8-8ef2-fac239546c3c) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/37120 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/32437 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/78267 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/108184 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/32890 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/137390 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/54296 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/41855 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/105681 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/54807 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/110702 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/105332 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26865 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/50851 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/29292 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/51895 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/54233 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/60409 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/53468 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/56924 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/55226 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/18cda55ff80977626c5854b56fcf4f1776b64dda. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51978

alanbaradlay commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -1496,7 +1496,7 @@ LayoutUnit RenderBlockFlow::collapseMarginsWithChildInfo(RenderBox* child, Margi
     // margin top of the self-collapsing block. If the resulting collapsed margin leaves the child still intruding into the float then we will want to clear it.
     if (!marginInfo.canCollapseWithMarginBefore()) {
-        if (auto value = selfCollapsingMarginBeforeWithClear(child->previousSibling()))
+        if (auto value = selfCollapsingMarginBeforeWithClear(child->previousInFlowSibling()))
```
how about floats? (they are also not inflow).

zakariaridouh committed:
[Layout] Fix crash in WebCore::RenderGrid::gridAreaRangeForOutOfFlow
https://bugs.webkit.org/show_bug.cgi?id=300354
rdar://161639703

Reviewed by Alan Baradlay.

Fix a crash in WebCore::RenderGrid::gridAreaRangeForOutOfFlow where we haven't completed
in-flow placement and grid sizing yet, and so our positions vector doesn't have valid values.

This aligns us closer with the spec, where only in flow children are
eligible for margin collapsing.

The root cause of this crash is we're attempting to do margin collapsing with an out of flow child, but
we can only do this for in flow children.

[1] <https://drafts.csswg.org/css2/#collapsing-margins>

* LayoutTests/fast/css-grid-layout/grid-crash-area-range-for-out-of-flow-expected.txt: Added.
* LayoutTests/fast/css-grid-layout/grid-crash-area-range-for-out-of-flow-no-text-expected.txt: Added.
* LayoutTests/fast/css-grid-layout/grid-crash-area-range-for-out-of-flow-no-text.html: Added.
* LayoutTests/fast/css-grid-layout/grid-crash-area-range-for-out-of-flow.html: Added.
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::isSelfCollapsingBlock const):

Canonical link: https://commits.webkit.org/301976@main
